File: Auto/final_code/python/Motor3Module.py
import RPi.GPIO as GPIO
from time import sleep

# Setze die GPIO-Pins
GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)

# Klasse zur Steuerung des Motors
class Motor():
    def __init__(self, EnaA, In1A, In2A, EnaB, In1B, In2B, EnaC, In1C, In2C):
        self.EnaA = EnaA
        self.In1A = In1A
        self.In2A = In2A
        self.EnaB = EnaB
        self.In1B = In1B
        self.In2B = In2B
        self.EnaC = EnaC
        self.In1C = In1C
        self.In2C = In2C
        
        # GPIO-Pins für die Motorsteuerung initialisieren
        GPIO.setup(self.EnaA, GPIO.OUT)
        GPIO.setup(self.In1A, GPIO.OUT)
        GPIO.setup(self.In2A, GPIO.OUT)
        GPIO.setup(self.EnaB, GPIO.OUT)
        GPIO.setup(self.In1B, GPIO.OUT)
        GPIO.setup(self.In2B, GPIO.OUT)
        GPIO.setup(self.EnaC, GPIO.OUT)
        GPIO.setup(self.In1C, GPIO.OUT)
        GPIO.setup(self.In2C, GPIO.OUT)
        
        # PWM-Objekte für die Motoren erstellen und starten
        self.pwmA = GPIO.PWM(self.EnaA, 100)
        self.pwmA.start(0)
        self.pwmB = GPIO.PWM(self.EnaB, 100)
        self.pwmB.start(0)
        self.pwmC = GPIO.PWM(self.EnaC, 100)
        self.pwmC.start(0)
 
    def move(self, speed=0.5, turn=0, t=0):
        # Geschwindigkeit und Lenkwinkel in PWM-Duty-Cycle-Werte umrechnen
        speed *= 100
        turn *= 100
 
        # PWM-Duty-Cycle für die Motoren setzen
        self.pwmA.ChangeDutyCycle(abs(speed))
        self.pwmB.ChangeDutyCycle(abs(speed))
        self.pwmC.ChangeDutyCycle(abs(turn))
 
        # Motorrichtungen entsprechend der Geschwindigkeit einstellen
        if speed > 0:
           GPIO.output(self.In1A, GPIO.HIGH)
           GPIO.output(self.In2A, GPIO.LOW)
           GPIO.output(self.In1B, GPIO.LOW)
           GPIO.output(self.In2B, GPIO.HIGH)
        elif speed < 0:
           GPIO.output(self.In1A, GPIO.LOW)
           GPIO.output(self.In2A, GPIO.HIGH)
           GPIO.output(self.In1B, GPIO.HIGH)
           GPIO.output(self.In2B, GPIO.LOW)

        # Lenkrichtung und Geschwindigkeit für den Lenkmotor einstellen
        if turn > 0:
           GPIO.output(self.In1C, GPIO.HIGH)
           GPIO.output(self.In2C, GPIO.LOW)
           self.pwmA.ChangeDutyCycle(abs(turn))
           GPIO.output(self.In1A, GPIO.HIGH)
           GPIO.output(self.In2A, GPIO.LOW)
           self.pwmB.ChangeDutyCycle(abs(turn))
           GPIO.output(self.In1B, GPIO.HIGH)
           GPIO.output(self.In2B, GPIO.LOW)
        elif turn < 0:
           GPIO.output(self.In1C, GPIO.LOW)
           GPIO.output(self.In2C, GPIO.HIGH)
           self.pwmA.ChangeDutyCycle(abs(turn))
           GPIO.output(self.In1A, GPIO.LOW)
           GPIO.output(self.In2A, GPIO.HIGH)
           self.pwmB.ChangeDutyCycle(abs(turn))
           GPIO.output(self.In1B, GPIO.LOW)
           GPIO.output(self.In2B, GPIO.HIGH)
 
        sleep(t)
        
    def stop(self, t=0):
        # Alle Motoren stoppen
        self.pwmA.ChangeDutyCycle(0)
        self.pwmB.ChangeDutyCycle(0)
        self.pwmC.ChangeDutyCycle(0)
        sleep(t)

def main():
    # Bewege den Motor vorwärts, lenke leicht und halte an
    motor.move(0.4, 0.2, 2)
    motor.stop(2)
    
    # Bewege den Motor rückwärts, lenke leicht und halte an
    motor.move(-0.3, -0.2, 2)
    motor.stop(2)

if __name__ == '__main__':
    # Initialisiere den Motor mit den entsprechenden GPIO-Pins
    motor = Motor(10, 9, 11, 17, 22, 27, 14, 15, 18)
    main()

File: Auto/Test_code/C++/C++ first edition/Motor3Module.cpp
#include <wiringPi.h>
#include <softPwm.h>
#include <unistd.h>

class Motor {
private:
    int EnaA, In1A, In2A, EnaB, In1B, In2B, EnaC, In1C, In2C;

public:
    Motor(int EnaA, int In1A, int In2A, int EnaB, int In1B, int In2B, int EnaC, int In1C, int In2C) {
        this->EnaA = EnaA;
        this->In1A = In1A;
        this->In2A = In2A;
        this->EnaB = EnaB;
        this->In1B = In1B;
        this->In2B = In2B;
        this->EnaC = EnaC;
        this->In1C = In1C;
        this->In2C = In2C;

        pinMode(this->EnaA, OUTPUT);
        pinMode(this->In1A, OUTPUT);
        pinMode(this->In2A, OUTPUT);
        pinMode(this->EnaB, OUTPUT);
        pinMode(this->In1B, OUTPUT);
        pinMode(this->In2B, OUTPUT);
        pinMode(this->EnaC, OUTPUT);
        pinMode(this->In1C, OUTPUT);
        pinMode(this->In2C, OUTPUT);

        softPwmCreate(this->EnaA, 0, 100);
        softPwmCreate(this->EnaB, 0, 100);
        softPwmCreate(this->EnaC, 0, 100);
    }

    void move(double speed = 0.5, double turn = 0, int t = 0) {
        int pwmSpeed = static_cast<int>(speed * 100);
        int pwmTurn = static_cast<int>(turn * 100);

        softPwmWrite(this->EnaA, abs(pwmSpeed));
        softPwmWrite(this->EnaB, abs(pwmSpeed));
        softPwmWrite(this->EnaC, abs(pwmTurn));

        if (pwmSpeed > 0) {
            digitalWrite(this->In1A, HIGH);
            digitalWrite(this->In2A, LOW);
            digitalWrite(this->In1B, LOW);
            digitalWrite(this->In2B, HIGH);
        } else if (pwmSpeed < 0) {
            digitalWrite(this->In1A, LOW);
            digitalWrite(this->In2A, HIGH);
            digitalWrite(this->In1B, HIGH);
            digitalWrite(this->In2B, LOW);
        }

        if (pwmTurn > 0) {
            digitalWrite(this->In1C, HIGH);
            digitalWrite(this->In2C, LOW);
            softPwmWrite(this->EnaA, abs(pwmTurn));
            digitalWrite(this->In1A, HIGH);
            digitalWrite(this->In2A, LOW);
            softPwmWrite(this->EnaB, abs(pwmTurn));
            digitalWrite(this->In1B, HIGH);
            digitalWrite(this->In2B, LOW);
        } else if (pwmTurn < 0) {
            digitalWrite(this->In1C, LOW);
            digitalWrite(this->In2C, HIGH);
            softPwmWrite(this->EnaA, abs(pwmTurn));
            digitalWrite(this->In1A, LOW);
            digitalWrite(this->In2A, HIGH);
            softPwmWrite(this->EnaB, abs(pwmTurn));
            digitalWrite(this->In1B, LOW);
            digitalWrite(this->In2B, HIGH);
        }

        sleep(t);
    }

    void stop(int t = 0) {
        softPwmWrite(this->EnaA, 0);
        softPwmWrite(this->EnaB, 0);
        softPwmWrite(this->EnaC, 0);
        sleep(t);
    }
};

// int main() {
//     wiringPiSetup();

//     Motor motor(10, 9, 11, 17, 22, 27, 14, 15, 18);

//     motor.move(0.4, 0.2, 2);
//     motor.stop(2);
//     motor.move(-0.3, -0.2, 2);
//     motor.stop(2);

//     return 0;
// }